enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/ForsetiClient.java

```diff
@@ -95,96 +105,124 @@ public ForsetiClient( int id,
         }
     }
 
+    /**
+     * Reset current client state. Make it ready for next bunch of operations.
+     * Should be used before factory release client to public usage.
+     */
+    public void reset()
+    {
+        closed = false;
+        activeOperations.set( 0 );
+    }
+
     @Override
     public void acquireShared( Locks.ResourceType resourceType, long resourceId ) throws AcquireLockTimeoutException
     {
-        // Grab the global lock map we will be using
-        ConcurrentMap<Long, ForsetiLockManager.Lock> lockMap = lockMaps[resourceType.typeId()];
-
-        // And grab our local lock maps
-        PrimitiveLongIntMap heldShareLocks = sharedLockCounts[resourceType.typeId()];
-        PrimitiveLongIntMap heldExclusiveLocks = exclusiveLockCounts[resourceType.typeId()];
-
-        // First, check if we already hold this as a shared lock
-        int heldCount = heldShareLocks.get( resourceId );
-        if(heldCount != -1)
+        // client closed exiting
+        if ( closed )
         {
-            // We already have a lock on this, just increment our local reference counter.
-            heldShareLocks.put( resourceId, heldCount + 1 );
             return;
         }
-
-        // Second, check if we hold it as an exclusive lock
-        if( heldExclusiveLocks.containsKey( resourceId ) )
+        try
         {
-            // We already have an exclusive lock, so just leave that in place. When the exclusive lock is released,
-            // it will be automatically downgraded to a shared lock, since we bumped the share lock reference count.
-            heldShareLocks.put( resourceId, 1 );
-            return;
-        }
+            activeOperations.incrementAndGet();
```

Comment:
Thx, sure!

Yes, it's very very true, miss the race there.

In cases when thread owning locks is doing something we can check graceful exit flag and that should not be a problem, but there are also cases when those threads belong to hanging clients and will not do anything anymore.